pkg/lib/collections/priority_queue.go

```diff
@@ -0,0 +1,205 @@
+)
+
+// PriorityQueue contains items of type T, and allows you to enqueue
+// and dequeue items with a specific priority. Items are dequeued in
+// highest priority first order.
+type PriorityQueue[T any] struct {
+	internalQueue queueHeap
+	mu            sync.Mutex
+}
+
+// MatchingFunction can be used when 'iterating' the priority queue to find
+// items with specific properties.
+type MatchingFunction[T any] func(possibleMatch T) bool
+
+// NewPriorityQueue creates a new ptr to a priority queue for type T.
+func NewPriorityQueue[T any]() *PriorityQueue[T] {
+	q := &PriorityQueue[T]{
+		internalQueue: make(queueHeap, 0, InitialQueueCapacity),
+	}
+	heap.Init(&q.internalQueue)
+	return q
+}
+
+// Enqueue will add the item specified by `data` to the queue with the
+// the priority given by `priority`.
+func (pq *PriorityQueue[T]) Enqueue(data T, priority int) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	pq.enqueue(data, priority)
+}
+
+// enqueue is a lock-free version of Enqueue for internal use when a
+// method already has a lock.
+func (pq *PriorityQueue[T]) enqueue(data T, priority int) {
+	heap.Push(
+		&pq.internalQueue,
+		&heapItem{
+			value:    data,
+			priority: priority,
+		},
+	)
+}
+
+// Dequeue returns the next highest priority item, returning both
+// the data Enqueued previously, and the priority with which it was
+// enqueued. An err (ErrEmptyQueue) may be returned if the queue is
+// currently empty.
+func (pq *PriorityQueue[T]) Dequeue() (T, int, error) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	return pq.dequeue()
+}
+
+// dequeue is a lock-free version of Dequeue() for use by internal
+// methods that already have a lock
+func (pq *PriorityQueue[T]) dequeue() (T, int, error) {
+	if pq.IsEmpty() {
+		return *new(T), 0, ErrEmptyQueue
+	}
+
+	internalItem := heap.Pop(&pq.internalQueue)
+	heapItem := internalItem.(*heapItem)
+	item, ok := heapItem.value.(T)
+	if !ok {
+		// Something has gone very, very wrong if the item in the heap
+		// is not a T as our Enqueue method should be the only thing
+		// putting stuff into it.
+		return item, 0, errors.New("priority queue found a bad type in the internal heap")
+	}
+
+	return item, heapItem.priority, nil
+}
+
+// DequeueWhere allows the caller to iterate through the queue, in priority order, and
+// attempt to match an item using the provided `MatchingFunction`.  This method has a high
+// time cost as dequeued but non-matching items must be held and requeued once the process
+// is complete.  Luckily, we use the same amount of space (bar a few bytes for the
+// extra PriorityQueue) for the dequeued items.
+func (pq *PriorityQueue[T]) DequeueWhere(matcher MatchingFunction[T]) (T, int, error) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	var result T
+	var priority int
+	var found bool
+
+	// Create a new Q to hold items that are not matches, this is suboptimal for time
+	// but not really an issue for space as we'll be using the same.
+	newQ := NewPriorityQueue[T]()
```

Comment:
it doesn't look like you need a temporary queue to hold the polled items as you are going to merge and sort them again when merging with the main queue. Unless merging sorted items is more efficient, but in this case the `newQ` should be in opposite order so you are merging the least priority item first. I might be wrong about this :S